File: clients/PHP-client/benchmark_parallel.php
<?php
require_once __DIR__ . '/vendor/autoload.php';
require_once __DIR__ . '/src/Client.php';
require_once __DIR__ . '/src/Exception/DriverException.php';

use SoliDB\Client;

function run_parallel_benchmark()
{
    $port = intval(getenv('SOLIDB_PORT') ?: '9998');
    $password = getenv('SOLIDB_PASSWORD') ?: 'password';

    $numWorkers = 16;
    $totalInserts = 10000;
    $insertsPerWorker = $totalInserts / $numWorkers;

    $db = 'bench_db';
    $col = 'php_parallel_bench';

    // Setup: create database and collection
    $setupClient = new Client('127.0.0.1', $port);
    $setupClient->auth('_system', 'admin', $password);

    try {
        $setupClient->createDatabase($db);
    } catch (\Exception $e) {
    }

    try {
        $setupClient->createCollection($db, $col);
    } catch (\Exception $e) {
    }

    $startTime = microtime(true);

    // PHP doesn't have true threading, but we can use pcntl_fork for processes
    // or parallel extension. For simplicity, we'll use sequential with batching
    // as a workaround since parallel is not always available

    if (function_exists('pcntl_fork')) {
        // Use forking for true parallelism
        $pids = [];

        for ($w = 0; $w < $numWorkers; $w++) {
            $pid = pcntl_fork();

            if ($pid == -1) {
                die("Could not fork");
            } elseif ($pid == 0) {
                // Child process
                $client = new Client('127.0.0.1', $port);
                $client->auth('_system', 'admin', $password);

                for ($i = 0; $i < $insertsPerWorker; $i++) {
                    $client->insert($db, $col, [
                        'worker' => $w,
                        'id' => $i,
                        'data' => 'parallel benchmark data'
                    ]);
                }
                exit(0);
            } else {
                $pids[] = $pid;
            }
        }

        // Wait for all children
        foreach ($pids as $pid) {
            pcntl_waitpid($pid, $status);
        }
    } else {
        // Fallback: sequential execution (no true parallelism)
        $client = new Client('127.0.0.1', $port);
        $client->auth('_system', 'admin', $password);

        for ($i = 0; $i < $totalInserts; $i++) {
            $client->insert($db, $col, [
                'id' => $i,
                'data' => 'parallel benchmark data'
            ]);
        }
    }

    $endTime = microtime(true);
    $duration = $endTime - $startTime;
    $opsPerSec = $totalInserts / $duration;

    echo "PHP_PARALLEL_BENCH_RESULT:" . round($opsPerSec, 2) . "\n";
}

run_parallel_benchmark();

File: clients/go-client/benchmark_parallel.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"sync"
	"time"

	"github.com/solisoft/solidb-go-client/solidb"
)

func main() {
	portStr := os.Getenv("SOLIDB_PORT")
	if portStr == "" {
		portStr = "9998"
	}
	port, _ := strconv.Atoi(portStr)
	password := os.Getenv("SOLIDB_PASSWORD")
	if password == "" {
		password = "password"
	}

	numWorkers := 16
	totalInserts := 10000
	insertsPerWorker := totalInserts / numWorkers

	db := "bench_db"
	col := "go_parallel_bench"

	// Setup: create database and collection
	setupClient := solidb.NewClient("127.0.0.1", port)
	setupClient.Connect()
	setupClient.Auth("_system", "admin", password)
	setupClient.CreateDatabase(db)
	setupClient.CreateCollection(db, col, nil)
	setupClient.Close()

	var wg sync.WaitGroup
	startTime := time.Now()

	for w := 0; w < numWorkers; w++ {
		wg.Add(1)
		go func(workerID int) {
			defer wg.Done()

			client := solidb.NewClient("127.0.0.1", port)
			err := client.Connect()
			if err != nil {
				fmt.Printf("Worker %d: connection error: %v\n", workerID, err)
				return
			}
			defer client.Close()

			client.Auth("_system", "admin", password)

			for i := 0; i < insertsPerWorker; i++ {
				data := map[string]interface{}{
					"worker": workerID,
					"id":     i,
					"data":   "parallel benchmark data",
				}
				client.Insert(db, col, data, nil)
			}
		}(w)
	}

	wg.Wait()
	duration := time.Since(startTime)

	opsPerSec := float64(totalInserts) / duration.Seconds()
	fmt.Printf("GO_PARALLEL_BENCH_RESULT:%.2f\n", opsPerSec)
}
